Overhaul settings panel layout

dragunoff opened the pull request:
This PR lays out the settings panel in a consistent and simple way whilst accommodating any number of any type of settings widgets. 

Closes #18952
Closes #18426
Related #16596

### RA, D2k, TS
The vertical tabs are a precedent but I think they fit and this layout allows for more tabs to be added. In the "common" UI I decided not to touch it as there isn't enough horizontal space because of the main menu. The settings panels is made larger and is the size of the asset browser.
![Screenshot_20210922_195713](https://user-images.githubusercontent.com/1355810/134388570-249ef9b9-c845-4dd4-a13c-2acf45ac8133.png)

### TD
In TD the vertical vs horizontal tabs dissonance was more apparent so I switched the game info to also use the new vertically stacked layout. And both panels are now the same size.
![Screenshot_20210922_172916](https://user-images.githubusercontent.com/1355810/134387618-6162874a-5330-4b98-b396-9584a0916f81.png)
Settings panel with vertical tabs

![Screenshot_20210922_172900](https://user-images.githubusercontent.com/1355810/134387617-7ecf98d0-5d7b-4453-81da-b074cac4ebd3.png)
Game info panel with vertical tabs

dragunoff committed:
Overhaul settings panel layout

- make the panel larger
- place settings widgets in a scroll panel
- arrange settings widgets in two columns
- make tabs in TD vertical

dragunoff committed:
Update TD game info panel to match the settings panel

pchote commented:
I generally like this, but there are a small number of sharp edges that we should try to polish:
* The group header inside the scroll panel for Audio and Input are redundant and IMO would be better if removed
* The sliders, dropdowns, and hotkeys look good in a single column, but I feel that most (all?) of the checkboxes would be better arranged with sharing two columns
* The control scheme description text feels awkward - maybe moving it right so it is clearly indented below the option would help?
* The frame limiter checkbox and slider are really awkward too (the slider modifies the checkbox that is above it)
* Can we please reduce the vertical stride of the TD tabs from 50 to 45? 50 feels noticeably wrong to me, 45 seems better and also matches the horizontal button separation we use in other places

dragunoff commented:
> 
> The group header inside the scroll panel for Audio and Input are redundant and IMO would be better if removed
> 

Agreed.

> The sliders, dropdowns, and hotkeys look good in a single column, but I feel that most (all?) of the checkboxes would be better arranged with sharing two columns
> 

I'm not against arranging controls in columns in principle but I think that our settings panels are too narrow for that. I'd stick with a single column.

> The control scheme description text feels awkward - maybe moving it right so it is clearly indented below the option would help?
> 

I think that it fits better to the left, putting it below the dropdown feels more awkward:
![Screenshot_20210928_111603](https://user-images.githubusercontent.com/1355810/135051109-82d503a4-a7e7-41ad-b664-6e256f0be13a.png)


> The frame limiter checkbox and slider are really awkward too (the slider modifies the checkbox that is above it)
> 

Yes, this bugs me too. I explored two approaches:

Enable/disable the silder:
https://user-images.githubusercontent.com/1355810/135052112-f385724d-7d06-4b1b-b028-03ae94a7bbcc.mp4

Show/hide the slider:
https://user-images.githubusercontent.com/1355810/135052130-7c97ed43-eb0b-44c7-8587-87697f9271bf.mp4

> Can we please reduce the vertical stride of the TD tabs from 50 to 45? 50 feels noticeably wrong to me, 45 seems better and also matches the horizontal button separation we use in other places

Agreed.

pchote commented:
The thing that bothers me with the checkboxes is that all the other input types focus their input/attention to the right of the panel, but checkboxes focuses it on the far left, making those tabs feels disjointed.

My first thought was that we could mitigate this by having two checkboxes per row, so at least we have some kind of content in the right half of those rows.

My second thought is that we could instead align the checkboxes to match the dropdowns and sliders:

<img width="733" alt="Screenshot 2021-09-28 at 09 58 53" src="https://user-images.githubusercontent.com/167819/135057668-5cd2f7d9-428b-4ac2-8f84-a722d00f0867.png">

This is only a problem where there are checkboxes sandwiched between dropdowns, so the advanced panel could keep its checkboxes left aligned. This screenshot also shows a bug: changing the UI Scale makes all the tabs collapse on themselves -  `DisplaySettingsLogic.RecalculateWidgetLayout` will need a hack to skip these like it does for badges.

A more radical option could be to place the labels to the left:

<img width="732" alt="Screenshot 2021-09-28 at 10 12 31" src="https://user-images.githubusercontent.com/167819/135058925-ad0ddc97-ec43-4ad9-a038-81303ab5c00f.png">

But i don't like how this leaves only a small visible click target to toggle the option.

dragunoff commented:
Alright, I see and I agree that the layout can be improved. How about right-aligning all labels:
![image](https://user-images.githubusercontent.com/1355810/135086631-68fa5887-0fdf-41a4-9550-4a92e6894d9c.png)
This has the benefit of bringing the labels closer to the controls. Checkboxes of this kind are a precedent in the OpenRA UI but I don't think they look too out of place. To improve the usability we can make the labels trigger the control that they are describing (like the `for` attribute in HTML).

An alternative is to make the panel wider, place all labels (except on checkboxes) above the controls and go all out 2-column. Roughly like this screen from BroodWar:
![image](https://user-images.githubusercontent.com/1355810/135088282-d4cbc865-bd3a-4da8-be8e-8d751133d486.png)

pchote commented:
The broodwar approach could potentially work well and IMO is definitely worth at least mockup in our common/TD UI. We could do the same for the lobby options bin, which would let us use 3 columns for dropdowns instead of mixing 2 and 3 columns in the UI.

dragunoff committed:
Update introductory prompts to match settings layout

dragunoff committed:
Use disabled button state for slider thumb

dragunoff committed:
Remove redundant default hotkey UI notice

dragunoff commented:
Layout changed to 2-columns:
![Screenshot_20210929_164708](https://user-images.githubusercontent.com/1355810/136077374-2785b491-7d6b-4cac-8ef4-69e2a7adf941.png)
![Screenshot_20210929_165000](https://user-images.githubusercontent.com/1355810/136077476-139fb0d0-8436-4d06-a9ef-7ea884b6e385.png)

Also updated the introductory prompts to match:
![Screenshot_20210929_205411](https://user-images.githubusercontent.com/1355810/136077488-fba6e57b-5399-43d4-a04b-e98384803e38.png)
![Screenshot_20210929_210730](https://user-images.githubusercontent.com/1355810/136077559-0479a1ae-84bc-498e-a1ff-da1cad3de177.png)

And a few other things:
- I added the tab container to the hack in `DisplaySettingsLogic.RecalculateWidgetLayout`.e
- Also fixed a bug with the disabled state of the slider thumb.
- And finally I removed the redundant "This is the default" message for hotkeys. If this is not desired I will drop it.

dragunoff commented:
I have some more tweaking to do here so please hold off reviewing for a little while.

dragunoff committed:
Update common game ingame info tabs to match settings layout

dragunoff committed:
Update lobby options tab layout to match settings layout

dragunoff commented:
This is now ready for review.

Updated common game ingame info tabs to match settings layout
![Screenshot-20211007112652-1282x807](https://user-images.githubusercontent.com/1355810/136348733-bf49d6c1-7ad0-457c-b394-14d6c99aadee.png)

Updated lobby options tab layout to match settings layout
![Screenshot-20211007112556-1282x807](https://user-images.githubusercontent.com/1355810/136348757-49c46a18-f0cb-430b-940c-4a94e1aba3c3.png)

dragunoff commented:
Update: changed how the settings widget widths and positions are calculated (in YAML). This doesn't alter the appearance, just makes things a little more robust.

anvilvapre commented:
Would there be room for a checkbox for `GameSettings.EnableDiscordService` #19343?

dragunoff commented:
> `EnableDiscordService`

Yes, the new layout can accommodate any number of settings widgets. A checkbox for discord seems like a sensible feature but should be added in a follow up PR.

dragunoff commented:
Rebased.

dragunoff commented:
Rebased after the merge of #19681 so it all comes together.

pchote commented:
The darker separator style in RA and (to a lesser extent) TS still feel weird to me, but I imagine i'll get used to them if they go ahead. I do object to these separators having a different style to the separators used in the mission and multiplayer browsers, though. Can we please switch those to the same style, or keep everything using the old style here and open a new PR to update the separators in all four places as a followup?

pchote commented:
Perhaps best in a followup PR, but IMO good polish to fix before a release: The labels for the lobby options dropdown buttons should end with a colon to match the settings menu, and be greyed out when disabled (in the ingame menu / for non-admins in the lobby).

pchote reviewed (commented):
LGTM otherwise

dragunoff commented:
> The darker separator style in RA and (to a lesser extent) TS still feel weird to me, but I imagine i'll get used to them if they go ahead. I do object to these separators having a different style to the separators used in the mission and multiplayer browsers, though. Can we please switch those to the same style, or keep everything using the old style here and open a new PR to update the separators in all four places as a followup?

Agreed. It bugs me too. Let's keep the old style here and defer the new style to a follow up.


> Perhaps best in a followup PR, but IMO good polish to fix before a release: The labels for the lobby options dropdown buttons should end with a colon to match the settings menu, and be greyed out when disabled (in the ingame menu / for non-admins in the lobby).

Good catch. I'll add the colons here but leave the greying out to a follow up as it's not so straight forward - we need to establish a link between the label and the dropdown.

Thanks for the review @pchote.

dragunoff commented:
Update: 
- switched the separator to the old style (`button`)
- added colons after the labels of lobby dropdowns

reaperrr commented on mods/common/chrome/settings-display.yaml:
```diff
@@ -1,224 +1,317 @@
+									Width: PARENT_RIGHT
+									Height: 20
+									Font: Regular
+									Text: UI Feedback in Transients Panel
```
@pchote not entirely sure, but shouldn't this be `Transient Panels`?

pchote replied:
I think Transients Panel is indeed correct, since this is referring to the new ingame display for transient feedback.

dragunoff replied:
It does sound somewhat unclear. Should we change to something like "Show UI Feedback Notifications"? Or add a tooltip description? (not necessarily in this PR)

pchote replied:
> Show UI Feedback Notifications

👍

reaperrr replied:
Sounds good, but I'll merge this first then (there are some other improvements I'd like to see as well, I'll file a follow-up PR or issue about it later).

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/949e19e2755d36c3db356703dbcdd5100c3af9e1...b03596d690670fe94ebb0f9fbb76a6ed00881bc6
